core-common/src/main/java/org/zalando/nakadi/domain/EnvelopeHolder.java

```diff
@@ -0,0 +1,91 @@
+package org.zalando.nakadi.domain;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+
+public class EnvelopeHolder {
+
+    private byte metadataVersion;
+    private int metadataLength;
+    private int payloadLength;
+    private byte[] data;
+
+    public interface EventWriter {
+        void write(OutputStream os) throws IOException;
+    }
+
+    public static byte[] produceBytes(
+            final byte metadataVersion,
+            final EventWriter metadataWriter,
+            final EventWriter eventWriter) throws IOException {
+        final ByteArrayOutputStream eventOutputStream = new ByteArrayOutputStream();
+        eventOutputStream.write(metadataVersion);
+        eventOutputStream.write(new byte[4]);
+        metadataWriter.write(eventOutputStream);
+        final int metadataLength = eventOutputStream.size() - 1 - 4;
+        eventOutputStream.write(new byte[4]);
```

Comment:
We could save 2 object allocations here if we just had a (static) 4-byte placeholder array.  Not that important, but rather a low-hanging fruit.